collector/btrfs_linux.go

```diff
@@ -99,39 +248,95 @@ func (c *btrfsCollector) updateBtrfsStats(ch chan<- prometheus.Metric, s *btrfs.
 
 		ch <- prometheus.MustNewConstMetric(
 			desc,
-			prometheus.GaugeValue,
+			m.metricType,
 			m.value,
 			labelValues...,
 		)
 	}
 }
 
 // getMetrics returns metrics for the given Btrfs statistics.
-func (c *btrfsCollector) getMetrics(s *btrfs.Stats) []btrfsMetric {
+func (c *btrfsCollector) getMetrics(s *btrfs.Stats, iocStats *ioctlFsStats) []btrfsMetric {
 	metrics := []btrfsMetric{
 		{
 			name:            "info",
 			desc:            "Filesystem information",
 			value:           1,
+			metricType:      prometheus.GaugeValue,
 			extraLabel:      []string{"label"},
 			extraLabelValue: []string{s.Label},
 		},
 		{
-			name:  "global_rsv_size_bytes",
-			desc:  "Size of global reserve.",
-			value: float64(s.Allocation.GlobalRsvSize),
+			name:       "global_rsv_size_bytes",
+			desc:       "Size of global reserve.",
+			metricType: prometheus.GaugeValue,
+			value:      float64(s.Allocation.GlobalRsvSize),
 		},
 	}
 
 	// Information about devices.
-	for n, dev := range s.Devices {
-		metrics = append(metrics, btrfsMetric{
-			name:            "device_size_bytes",
-			desc:            "Size of a device that is part of the filesystem.",
-			value:           float64(dev.Size),
-			extraLabel:      []string{"device"},
-			extraLabelValue: []string{n},
-		})
+	if iocStats == nil {
+		for n, dev := range s.Devices {
+			metrics = append(metrics, btrfsMetric{
+				name:            "device_size_bytes",
+				desc:            "Size of a device that is part of the filesystem.",
+				metricType:      prometheus.GaugeValue,
+				value:           float64(dev.Size),
+				extraLabel:      []string{"device"},
+				extraLabelValue: []string{n},
+			})
+		}
+	} else {
+		for _, dev := range iocStats.devices {
+			// trim the /dev/ prefix from the device name so the value should match
+			// the value used in the fallback branch above
+			device := strings.TrimPrefix(dev.path, "/dev/")
```

Comment:
We need to handle non-default locations here. When running in container, the path might be /rootfs/dev or something

Reply:
Hmm, good spot! Are you happy `path.Split` is good enough?